### ISSUE TITLE
hbios - z180 real time unix time - (enhancement)

### ISSUE THREAD

**Issue by feilipu** [wwarthen/RomWBW#53]
I'd like to interface the HBIOS TIMER functions with the [Y2K](http://www.nongnu.org/avr-libc/user-manual/group__avr__time.html) epoch (related to Unix Epoch). Whilst it is possible as stands, would it be possible to suggest some minor changes to the HBIOS TICKS implementation (on Z180)? These suggestions would make things much simpler.

If the suggestion is approved, I'll do a PR to suit.

1. The Y2K Epoch is 32 bits, with the tick increment per second. -> Make the timer location 4+1 bytes, with the lowest byte measuring fractional seconds. Unix Epoch can then be loaded from RTC into the 4 bytes second counter with standard conversion code.

2. Match the `PRT0` tick rate to a binary multiple (i.e. 256Hz or 128Hz) and have the fractional second (1 byte) counter roll over to increment the second counter (4 bytes). -> Keeps the interrupt handling very simple (as it is now). Keep translation of [fractional seconds to nanoeconds](https://github.com/z88dk/z88dk/blob/master/libsrc/_DEVELOPMENT/target/scz180/driver/time/z180/clock_gettime.asm) out of time critical code.

The outcome is to allow the z88dk [`sys/time.h`](https://github.com/z88dk/z88dk/blob/master/include/_DEVELOPMENT/sdcc/sys/time.h) unix like functions to easily connect into the HBIOS API.

Thoughts?

**Comment by wwarthen** [comment]
Hi Phillip,

I do like the idea of supporting a "seconds" counter.  There are many
instances where this would be useful.  However, the existing 50Hz tick
counter is very specifically intended to be optimal for MP/M and audio
playback.

I think it would be pretty straightforward to maintain a separate 32-bit
counter that just counts seconds.  Essentially, the seconds counter would
be incremented 50 ticks.  I would need to add something to set and query
the seconds counter, but that is not hard.  There would be very little
overhead to maintaining the additional counter because 49 interrupts out of
50 would do almost nothing extra.

Would this be a reasonable compromise?

By the way, I want to make sure you know that not all platforms running
RomWBW have timer hardware.

Thanks,

Wayne

**Comment by feilipu** [comment]
Wayne,

yes, if 50Hz is a critical timer then it makes a very good compromise to have a 50Hz fractional second counter feeding into a 4 byte Y2K Epoch second counter. 👍 

In fact the only compromise is finding a low cost (ticks) way to roll-over the 50Hz counter.

The rest of the `sys/timer.h` functionality is easily generated with any sub-second resolution and is no effort provided a seconds timer (uint32_t) and fractional second second (uint8_t) are available at well known locations.

Perhaps you can (or do) have an `.inc` file with addresses of fixed register locations in the ROMWBW shim API? That would make it dead easy.

I don't know if there's anything smarter than `cp 50` `jr NC,_seconds_handler` `ret` to test and roll-over a 50Hz counter?  But, probably the system impact of manual roll-over on 50Hz interrupt is lower than counting 256Hz with automatic roll-over, considering handling 4x more interrupts. So that doesn't need to be over-cooked. 

Yes, aware that Timers are not available everywhere. I'm really just interested in the z180 hardware solution, given that we have a PRT available. I'm keen to have a seconds counter can be updated from a RTC (if available) on boot, or by user input from the monitor or application.

I've become used to having Y2K (aka Arduino) epoch available, given that there is good code [available from Michael Duane Rice](https://github.com/feilipu/z88dk-libraries/tree/master/time) to generate correct dates and print broken down time.

IIRC, I also have some code to convert from DS1307 to Y2K epoch.... looking ah, yes. [here](https://github.com/feilipu/avrfreertos/blob/master/freeRTOS10xx/lib_rtc/rtc.c).

Leave it to you then, for contemplation?

Cheers, Phillip

**Comment by wwarthen** [comment]
Hi Phillip,

I just checked in some changes that I hope will handle your use case.  HBIOS now keeps a runnings "seconds" register in addition to the "ticks" register.  Accessing the seconds register is identical to the  ticks register.  The specific function codes are documented in the API.txt file in Source\HBIOS and also in "RomWBW Architecture.pdf".

The seconds register starts counting from zero at startup.  If you want to align it with the Unix epoch, then you will need to do that at OS startup.  The value can be set via the HBIOS API.

Let me know if this works for you.

-Wayne

**Comment by feilipu** [comment]
Hi Wayne,

thanks looks very good. I'll have some time this weekend to test started.

There are two things that would make it perfect for providing the time functions I'm looking at.
- Access to the 50ms counter byte `HB_SECTCK`, as it will become misaligned with the 32bit tick timer.
- Return the whole time `HB_SECS` and `HB_SECTCK` (including the 50ms counter) byte in one intrinsic function to allow a `get_time()` function to capture that _special_ moment. i.e. capture all time bytes between `HB_DI` and `HB_EI` functions, and return them with one call.

It matters not how those 5 bytes come out, nor that `HB_SECTCK` is counting backwards to zero, everything else can be managed by the calling function.

Thanks again.
Phillip

**Comment by feilipu** [comment]
Wayne,
I should also offer by way of explanation, that the [Duane Rice code](https://github.com/feilipu/z88dk-libraries/tree/master/time) doesn't need to have a millisecond timer. It only handles whole seconds. The use case for the whole of time is for the [`sys/time.h`](https://github.com/z88dk/z88dk/blob/master/include/_DEVELOPMENT/sdcc/sys/time.h) functions, that provide [_seconds.nanoseconds_](https://github.com/z88dk/z88dk/blob/master/libsrc/_DEVELOPMENT/target/scz180/driver/time/z180/clock_gettime.asm), with a [_0.nanoseconds_ resolution](https://github.com/z88dk/z88dk/blob/master/libsrc/_DEVELOPMENT/target/scz180/driver/time/z180/clock_getres.asm) provided.

At the moment, these two function examples are incorrect, but you can see what I'm aiming to provide. When I've done the integration, I'll edit this comment.

**Comment by wwarthen** [comment]
I think I understand.  If I simply return the SECTCK byte in register C in
the "get seconds" call, that would do it, right?  I will subtract the value
from 50 so that it is not inverted when returned by the API.

-Wayne

**Comment by feilipu** [comment]
Wayne,

something like this would work. Thoughts?

``` asm
ld hl,HB_SECTCK
ld a,49
HB_DI
sub (hl)
inc hl ; now pointing to HB_SECS
ld c,(hl)
inc hl
ld b,(hl)
inc hl
ld e,(hl)
inc hl
ld d,(hl)
HB_EI
ld l,c
ld h,b
ld c,a
ret
```
Thanks, Phillip

**Comment by wwarthen** [comment]
All excellent suggestions.  I will pursue them later today if I can find a bit of time.

However, I am pretty sure that initializing the SECTCK counter to 50 is correct.  Although there are 51 numbers between 0 and 50, there are only 50 "transitions".  Not sure if that is a good explanation.  Regardless, I will definitely review the logic.

Thank you very much for these suggestions!

**Comment by wwarthen** [comment]
Hi Phillip,

I reviewed the code and I am pretty sure that setting SECTCK to 50 is correct.  Can you double check this and let me know what you think?  I will merge your push request as soon as we resolve this.

Thanks,

Wayne

**Comment by feilipu** [comment]
Hi Wayne,

Yes you're right. 

But, just before noon, is it 11.60am? or is it 11.59am? Even though there are 60 minutes in an hour, you never have the 60th minute.

Same applies here the reported HB_SECTCK counter has to be reported as 49 ticks plus a full period of the PRT timer. And, in that case we also need to subtract 50 ticks per second to make sure that zero is reported (when it is during the first period).

I'll fix the PR now.

Phillip

**Comment by wwarthen** [comment]
Yes, you are right about PRT0.  Thanks for that.

I will merge your changes tomorrow along with some other stuff I have been working on.

Thanks!

**Comment by wwarthen** [comment]
OK, your changes are merged.

I did find a small issue with the PRT0 setup code in my testing.  Your changes rely on the static CPU speed variable.  However, most RomWBW platforms will dynamically determine the CPU speed and adjust as needed.  While I could do all the math to determine the PRT0 setup on the fly, it is a non-trivial and there are multiple places where reload counters are setup.  In the end, I went back to using a fixed 50Hz timer.  However, I did retain the TICKSPERSEC constant which is definitely useful.  I was also careful to retain the -1 offset when loading the reload counter.

Thanks,

Wayne
